### ISSUE TITLE
Introduce lap times in the database

### ISSUE THREAD

**Issue by enrique-lozano** [f1db/f1db#8]
This info can be extracted from the Ergast API: http://ergast.com/mrd/methods/laps/ (avalaible from 1996).

We also have the info in CSV format:
- https://www.kaggle.com/datasets/rohanrao/formula-1-world-championship-1950-2020?select=lap_times.csv
- http://ergast.com/mrd/db/#csv

**Comment by marceloverdijk** [comment]
It would be great to have lap times included in the F1DB.
Even if it's only from a certain year upwards like 1996.

The questions I have:

- How can we verify the ergast data? What is the source? Is there another source to compare this with? E.g. wrong or even missing data.
- How to easily obtain new lap times asap after a new race is finished?

**Comment by enrique-lozano** [comment]
The Ergast API seems reliable, and has been active for a few years I think (I'm not very sure about its sources).

Official lap time data can be found in FIA documents, for example:
In https://www.fia.com/events/fia-formula-one-world-championship/season-2022/saudi-arabian-grand-prix/eventtiming The PDF called Lap Analysis. I've been searching a lot and I can not find any other sources

I find the FIA page very difficult to carry out searches by race, and even more difficult to obtain the data (since it is in pdf, it is more complicated to scrape the data). I also think that the FIA has only uploaded documents to his page since 2019, I don't know where Ergast gets the rest of the information (there is a forum for questions, you can still ask).

So yes, it would be very very difficult and tedious to add the lap time information manually. Not to mention that it would undoubtedly be the largest table in the database.

**Comment by marceloverdijk** [comment]
It would indeed increase the size a lot.
Would you be willing to find out (via their forum) what their source is? 

I'm a bit hesitant to depend on them for new data as it might delay F1DB releases.

**Comment by enrique-lozano** [comment]
Hello, I have already left a comment on the forum, it seems that the admin takes a few days to respond sometimes. You can follow it here: http://ergast.com/mrd/feedback/comment-page-8/#comment-464157

In the event that this will greatly delay the data update after each race, perhaps an annual upload of this data could be made, or for now simply add this historical data and that's it.

**Comment by marceloverdijk** [comment]
Yes, it could be a separate dist not being released directly after every race.
But for future sustainability it would be better to know what Ergast uses as source.

**Comment by enrique-lozano** [comment]
Completely agree with you

**Comment by marceloverdijk** [comment]
Thanks for the update.
Although the main source is the FIA, is the data publicly available as reference?

**Comment by enrique-lozano** [comment]
I haven't found much about it. As I said, the FIA data seems to have been uploaded on the web only since 2019. Investigating a bit on the Internet you can find pages like this: https://pitwall.app/analysis/compare-lap-times which coincidentally also has the information since 1996 as Ergast. Beyond that, nothing more.

**Comment by marceloverdijk** [comment]
Yes I guess they copy from each other. But question is where to retrieve new values from, like after a race has finished. What is the source then...

**Comment by AntonioSchez32** [comment]
You can retrieve lap times since 2018 to present from Fast F1 API. They are using F1 live timing, so telemetry is pretty accurate. Although they are debugging errors that may happen due to red flags, et cetera

I cannot confirm you where Ergast lap times are retrieved from since 1996 to 2017. There may be some errors, but I think it is very reliable. Fast F1 is planning to take over Ergast since is being deprecated at the end of 2024.

**Comment by AntonioSchez32** [comment]
Hi,

I have used your database SQLite for a project in Power BI and the following files in the ZIP. There are four files: f1_drivers_original.csv, f1_drivers.csv, f1_race_entries.csv and f1_lap_times.csv. 

The lap times are only available for Grand Prix races, not for sprint races. This data has been retrieved from formula1db.com website with web scraping. 

[FORMULA 1 DB.zip](https://github.com/user-attachments/files/18181092/FORMULA.1.DB.zip)

In order to use the data, I have joined drivers_original (Name of the driver in the website) with drivers (Name of the driver in f1db) and after that with entries and lap times, so we can have available the driver_id from the drivers table. Finally, I have brought the constructor_id, tyre_manufacturer_id and engine_manufacturer_id joining with race_results using (race_id, driver_number, driver_id). Maybe it can be done easier, but it was my approach.

Thanks for everything!
